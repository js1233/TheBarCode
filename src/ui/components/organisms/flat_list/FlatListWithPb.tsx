import NoRecordFound from "assets/images/tbc.svg";
import { COLORS, FONT_SIZE, SPACE, STRINGS } from "config";
import { usePreferredTheme } from "hooks";
import React, { useCallback, useState } from "react";
import {
  ActivityIndicator,
  FlatList,
  FlatListProps,
  StyleProp,
  StyleSheet,
  View,
  ViewStyle
} from "react-native";
import {
  AppLabel,
  TEXT_TYPE
} from "ui/components/atoms/app_label/AppLabel";
import { LoadMore } from "ui/components/atoms/app_load_more/LoadMore";
import ErrorWithRetryView from "ui/components/molecules/ErrorWithRetryView";

export interface FlatListWithPbProps<ItemT> extends FlatListProps<ItemT> {
  shouldShowProgressBar?: boolean;
  error?: string;
  errorView?: (errorText: string) => React.ReactElement | null;
  retryCallback?: () => void;
  pullToRefreshCallback?: (onComplete: () => void) => void;
  style?: StyleProp<ViewStyle>;
  ItemSeparatorHeaderAndFooterComponent?: React.ComponentType<any> | null;
  isAllDataLoaded?: boolean;
  noRecordFoundText?: string;
  listRef?: React.RefObject<FlatList>;
  noRecordFoundImage?: React.ReactElement | null;
  noRecordFoundStyle?: StyleProp<ViewStyle>;
}

type Props<ItemT> = FlatListWithPbProps<ItemT>;

export function FlatListWithPb<ItemT extends any>(props: Props<ItemT>) {
  const {
    style,
    shouldShowProgressBar,
    error,
    errorView,
    retryCallback,
    pullToRefreshCallback,
    ItemSeparatorHeaderAndFooterComponent,
    isAllDataLoaded = true,
    noRecordFoundText = STRINGS.common.no_record_found,
    onEndReached,
    listRef,
    noRecordFoundImage,
    noRecordFoundStyle,
    ...rest
  } = props;

  const [refreshing, setRefreshing] = useState(false);
  const theme = usePreferredTheme();
  const onRefresh = useCallback(async () => {
    setRefreshing(true);
    pullToRefreshCallback?.(() => {
      setRefreshing(false);
    });
  }, [pullToRefreshCallback]);

  function shouldShowError() {
    return error !== undefined && !dataHasRecords();
  }

  function getErrorView() {
    if (error !== undefined && errorView !== undefined) {
      return errorView(error);
    } else {
      return (
        <ErrorWithRetryView
          text={error}
          retryCallback={retryCallback}
          style={styles.errorView}
        />
      );
    }
  }

  const { data } = rest;

  function dataHasRecords() {
    return (data?.length ?? 0) > 0;
  }

  const footerWrapper = React.memo<Props<any>>(() => {
    return (
      <>
        {!isAllDataLoaded && dataHasRecords() && (
          <View style={styles.loadMore}>
            <LoadMore testID="loader" />
          </View>
        )}
        {ItemSeparatorHeaderAndFooterComponent}
      </>
    );
  });

  function getNormalView() {
    if (shouldShowError()) {
      return getErrorView();
    }

    if (!shouldShowError()) {
      let ui = (
        <FlatList
          // initialNumToRender={7}
          ref={listRef}
          onEndReachedThreshold={1}
          refreshing={refreshing}
          onRefresh={
            pullToRefreshCallback === undefined ? undefined : onRefresh
          }
          onEndReached={
            isAllDataLoaded || !dataHasRecords() ? undefined : onEndReached
          }
          ItemSeparatorComponent={ItemSeparatorHeaderAndFooterComponent}
          ListHeaderComponent={ItemSeparatorHeaderAndFooterComponent}
          ListFooterComponent={footerWrapper}
          {...rest}
        />
      );

      if (!dataHasRecords()) {
        //No record found text visibility issue
        return (
          <>
            <View style={[styles.noRecordParent, noRecordFoundStyle]}>
              {noRecordFoundImage ? (
                noRecordFoundImage
              ) : (
                <NoRecordFound
                  width={"70%"}
                  height={"15%"}
                  fill={theme.themedColors.primaryColor}
                />
              )}

              <AppLabel
                text={noRecordFoundText}
                numberOfLines={0}
                textType={TEXT_TYPE.SEMI_BOLD}
                style={styles.noRecordFoundText}
              />
            </View>
            {ui}
          </>
        );
      } else {
        return ui;
      }
    }
  }

  return (
    <View style={style}>
      {shouldShowProgressBar && data === undefined && (
        <ActivityIndicator
          testID="initial-loader"
          size="large"
          color={theme.themedColors.interface["900"]}
          style={[
            styles.initialPb,
            { backgroundColor: COLORS.theme?.interface[50] }
          ]}
        />
      )}

      {!(shouldShowProgressBar && data === undefined) && getNormalView()}
    </View>
  );
}

const styles = StyleSheet.create({
  errorView: {
    width: "100%",
    height: "100%",
    justifyContent: "center",
    alignItems: "center",
    padding: SPACE.lg
  },
  noRecordParent: {
    width: "100%",
    height: "100%",
    justifyContent: "center",
    alignItems: "center",
    position: "absolute"
  },
  noRecord: {
    textAlign: "center",
    fontSize: FONT_SIZE.lg
  },
  loadMorePb: {
    width: "100%",
    justifyContent: "center",
    alignItems: "center",
    marginTop: 16,
    marginBottom: 16
  },
  initialPb: {
    width: "100%",
    height: "100%",
    justifyContent: "center",
    alignItems: "center"
  },
  loadMore: {
    marginTop: SPACE.lg,
    height: SPACE._3xl,
    justifyContent: "center",
    flexDirection: "column",
    alignItems: "center"
  },
  noRecordFoundText: {
    fontSize: FONT_SIZE.lg,
    padding: SPACE.md,
    textAlign: "center"
  }
});
