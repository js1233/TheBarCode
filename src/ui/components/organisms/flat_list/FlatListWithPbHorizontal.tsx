import { COLORS, FONT_SIZE, SPACE, STRINGS } from "config";
import { usePreferredTheme } from "hooks";
import React, { useCallback, useState } from "react";
import {
  ActivityIndicator,
  FlatList,
  FlatListProps,
  Image,
  StyleProp,
  StyleSheet,
  View,
  ViewStyle
} from "react-native";
import ErrorWithRetryView from "ui/components/molecules/ErrorWithRetryView";
import { shadowStyleProps } from "utils/Util";
import { AppLabel } from "ui/components/atoms/app_label/AppLabel";

interface OwnProps<ItemT> extends FlatListProps<ItemT> {
  shouldShowProgressBar?: boolean;
  error?: string;
  errorView?: (errorText: string) => React.ReactElement | null;
  retryCallback?: () => void;
  pullToRefreshCallback?: (onComplete: () => void) => void;
  style?: StyleProp<ViewStyle>;
  ItemSeparatorHeaderAndFooterComponent?: React.ComponentType<any> | null;
  isAllDataLoaded?: boolean;
  noRecordFoundText?: string;
  listRef?: React.RefObject<FlatList>;
  noRecordFoundImage?: React.ReactElement;
  noRecordFoundStyle?: StyleProp<ViewStyle>;
}

type Props<ItemT> = OwnProps<ItemT>;

export function FlatListWithPbHorizontal<ItemT extends any>(
  props: Props<ItemT>
) {
  const {
    style,
    shouldShowProgressBar,
    error,
    errorView,
    retryCallback,
    pullToRefreshCallback,
    ItemSeparatorHeaderAndFooterComponent,
    isAllDataLoaded = true,
    noRecordFoundText = STRINGS.common.no_record_found,
    onEndReached,
    listRef,
    noRecordFoundImage,
    noRecordFoundStyle,
    ...rest
  } = props;

  const { themedColors } = usePreferredTheme();
  const [refreshing, setRefreshing] = useState(false);
  const onRefresh = useCallback(async () => {
    setRefreshing(true);
    pullToRefreshCallback?.(() => {
      setRefreshing(false);
    });
  }, [pullToRefreshCallback]);

  function shouldShowError() {
    return error !== undefined && !dataHasRecords();
  }

  function getErrorView() {
    if (error !== undefined && errorView !== undefined) {
      return errorView(error);
    } else {
      return (
        <ErrorWithRetryView
          text={error}
          retryCallback={retryCallback}
          style={styles.errorView}
        />
      );
    }
  }

  const { data } = rest;

  function dataHasRecords() {
    return (data?.length ?? 0) > 0;
  }

  const footerWrapper = React.memo<Props<any>>(() => {
    return (
      <>
        {!isAllDataLoaded && dataHasRecords() && (
          <View style={styles.loadMore}>
            <ActivityIndicator
              size={30}
              color={themedColors.primaryColor}
            />
          </View>
        )}
        {ItemSeparatorHeaderAndFooterComponent}
      </>
    );
  });

  function getNormalView() {
    if (shouldShowError()) {
      return getErrorView();
    }

    if (!shouldShowError()) {
      let ui = (
        <FlatList
          // initialNumToRender={7}
          ref={listRef}
          onEndReachedThreshold={1}
          refreshing={refreshing}
          onRefresh={
            pullToRefreshCallback === undefined ? undefined : onRefresh
          }
          onEndReached={
            isAllDataLoaded || !dataHasRecords() ? undefined : onEndReached
          }
          ItemSeparatorComponent={ItemSeparatorHeaderAndFooterComponent}
          ListHeaderComponent={ItemSeparatorHeaderAndFooterComponent}
          ListFooterComponent={footerWrapper}
          {...rest}
          horizontal={true}
        />
      );

      if (!dataHasRecords()) {
        //No record found text visibility issue
        return (
          <>
            <View style={[styles.noRecordParent, noRecordFoundStyle]}>
              {noRecordFoundImage ? noRecordFoundImage : null}
              <AppLabel text={noRecordFoundText} />
            </View>
            {ui}
          </>
        );
      } else {
        return ui;
      }
    }
  }

  return (
    <View style={style}>
      {shouldShowProgressBar && data === undefined && (
        <View style={styles.loaderContainer}>
          <Image
            testID="initial-loader"
            source={require("assets/images/sample.gif")}
            style={{ width: 20, height: 20 }}
          />
        </View>
      )}

      {!(shouldShowProgressBar && data === undefined) && getNormalView()}
    </View>
  );
}

const styles = StyleSheet.create({
  errorView: {
    width: "100%",
    height: "100%",
    justifyContent: "center",
    alignItems: "center"
  },
  noRecordParent: {
    width: "100%",
    height: "100%",
    justifyContent: "center",
    alignItems: "center",
    position: "absolute",
    bottom: SPACE.lg
  },
  noRecord: {
    textAlign: "center",
    fontSize: FONT_SIZE.lg
  },
  loadMorePb: {
    width: "100%",
    justifyContent: "center",
    alignItems: "center",
    marginTop: 16,
    marginBottom: 16
  },
  initialPb: {
    width: "100%",
    height: "100%",
    justifyContent: "center",
    alignItems: "center"
  },
  loadMore: {
    height: SPACE._3xl,
    justifyContent: "center",
    flexDirection: "column",
    alignItems: "center"
  },
  loaderContainer: {
    justifyContent: "center",
    alignItems: "center",
    backgroundColor: COLORS.white,
    width: 30,
    height: 30,
    borderRadius: 20,
    ...shadowStyleProps
  }
});
