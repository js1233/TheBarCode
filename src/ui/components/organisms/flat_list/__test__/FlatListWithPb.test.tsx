import { render } from "@testing-library/react-native";
import React from "react";
import "react-native";
import { Text, TouchableOpacity } from "react-native";
import { FlatListWithPb } from "ui/components/organisms/flat_list/FlatListWithPb";

type mock = {
  id: string;
  name: string;
};

const mockList: mock[] = [
  {
    id: "id1",
    name: "name1"
  },
  {
    id: "id2",
    name: "name2"
  },
  {
    id: "id3",
    name: "name3"
  }
];

const keyExtractor = (item: mock) => item.id.toString();

const listItem = ({ item }: { item: mock }) => (
  <TouchableOpacity testID={`item-${item.id}`}>
    <Text>{item.name}</Text>
  </TouchableOpacity>
);

test("snapshot", () => {
  const rendered = render(
    <FlatListWithPb<mock>
      keyExtractor={keyExtractor}
      shouldShowProgressBar={true}
      renderItem={listItem}
      data={mockList}
      style={{
        flexGrow: 1,
        flexBasis: 0
      }}
    />
  ).toJSON();
  expect(rendered).toMatchSnapshot();
});

test("should contain provided list element", () => {
  const { queryByTestId } = render(
    <FlatListWithPb<mock>
      keyExtractor={keyExtractor}
      shouldShowProgressBar={true}
      renderItem={listItem}
      data={mockList}
      style={{
        flexGrow: 1,
        flexBasis: 0
      }}
    />
  );
  expect(queryByTestId(`item-${mockList[0].id}`)).not.toBeNull();
  expect(queryByTestId(`item-${mockList[1].id}`)).not.toBeNull();
  expect(queryByTestId(`item-${mockList[2].id}`)).not.toBeNull();
  expect(queryByTestId("item-id0")).toBeNull();
});

test("show progress bar before loading the list", async () => {
  const { queryByTestId } = render(
    <FlatListWithPb<mock>
      keyExtractor={keyExtractor}
      shouldShowProgressBar={true}
      renderItem={listItem}
      data={undefined}
      style={{
        flexGrow: 1,
        flexBasis: 0
      }}
    />
  );
  const loader = queryByTestId("initial-loader");
  expect(loader).not.toBeNull();
});

test("hide progress bar after loading the list", async () => {
  const { queryByTestId } = render(
    <FlatListWithPb<mock>
      keyExtractor={keyExtractor}
      shouldShowProgressBar={false}
      renderItem={listItem}
      data={mockList}
      style={{
        flexGrow: 1,
        flexBasis: 0
      }}
    />
  );
  const loader = queryByTestId("initial-loader");
  expect(loader).toBeNull();
});

test("show error view", async () => {
  const { queryAllByTestId } = render(
    <FlatListWithPb<mock>
      error="Something went wrong"
      keyExtractor={keyExtractor}
      shouldShowProgressBar={false}
      renderItem={listItem}
      data={mockList}
      style={{
        flexGrow: 1,
        flexBasis: 0
      }}
    />
  );
  const error = queryAllByTestId("error");
  expect(error).not.toBeNull();
});
