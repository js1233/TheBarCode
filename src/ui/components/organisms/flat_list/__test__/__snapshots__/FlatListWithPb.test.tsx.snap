// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<View
  style={
    Object {
      "flexBasis": 0,
      "flexGrow": 1,
    }
  }
>
  <RCTScrollView
    ListFooterComponent={
      Object {
        "$$typeof": Symbol(react.memo),
        "compare": null,
        "type": [Function],
      }
    }
    data={
      Array [
        Object {
          "id": "id1",
          "name": "name1",
        },
        Object {
          "id": "id2",
          "name": "name2",
        },
        Object {
          "id": "id3",
          "name": "name3",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={1}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshing={false}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={false}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="item-id1"
        >
          <Text>
            name1
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={false}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="item-id2"
        >
          <Text>
            name2
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          accessible={true}
          collapsable={false}
          focusable={false}
          nativeID="animatedComponent"
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="item-id3"
        >
          <Text>
            name3
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
      />
    </View>
  </RCTScrollView>
</View>
`;
